proposals/lotus-dynamic-retrieval-pricing.md

```diff
@@ -0,0 +1,142 @@
+and other factors.
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+
+Miners should be able to configure a flexible built-in policy (which enables the
+Bedrock golden path), as well as provide an external script that receives the
+decision factors wrapped in a structured message, and returns the pricing
+decisions. This could be similar to the existing deal filter mechanism.
+
+#### Impact
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+It's essential.
+
+#### Internal leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+Low. This is a chore, not a novelty seeking nor risky project.
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+10. This solves an existing shortcoming in our technology that we have "launch
+data" for.
+
+## Project definition
+
+![Solution sketch](./sketches/dynamic-retrieval-pricing.png)
+
+#### Brief plan of attack
+
+1. Introduce the ability to query the unsealing status of a piece from the
+   storage subsystem (unseal_status).
+2. Introduce the ability to query whether the storage deal was a verified deal
+   (verified_deal).
+3. Introduce the ability to query whether the storage deal had fast retrieval
+   enabled (fast_retrieval).
+4. Add the ability to set a price function:
+    - Built-in function (shipping with Lotus), which supports file
+      configuration. By default, it prices verified, fast-retrieval, unsealed
+      pieces at 0 (using the fields above).
+    - Via an externally invoked function, so that miners can inject their own
+      logic to quote dynamic prices -- similar to deal filter. This requires
+      defining a schema and an intermediate data format (likely JSON).
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+Merged pull requests on go-fil-markets and lotus + user documentation.
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+If the goals in the project brief are satisfied, we know we did the right thing.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+N/A.
+
+#### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+The pricing function could be hardcoded, but that's too short-sighted.
+
+#### Dependencies/prerequisites
+<!--List any other projects that are dependencies/prerequisites for this project that is being pitched.-->
+
+F3 retrieval stabilisation.
```

Comment:
I don't think this is really a pre-requisite - the ongoing stabilization work is for the case where the connection goes down in the middle of a data transfer, but a lot of retrieval deals will likely be made over stable connections.

Reply:
Agree. I'm going to refine this to say that it's not technically a dependency, but work here needs to be coordinated with F3 as they are touching the same codebases.